TensorFlow/freeze_tf2_model.py

```diff
@@ -1,140 +1,238 @@
+        #
+        # full_model = full_model.get_concrete_function(
+        #    (tf.TensorSpec(shape=None, dtype=tf.float32),
+        #    tf.TensorSpec(shape=None, dtype=tf.int64),
+        #    ...
+        #    ))
+
+        # Get frozen graph def
+        frozen_func = convert_variables_to_constants_v2(full_model)
+        frozen_func.graph.as_graph_def()
+
+        print('Saving frozen model to: ' + os.path.join(frozen_out_path, frozen_graph_filename + '.pb'))
+        tf.io.write_graph(graph_or_graph_def=frozen_func.graph,
+                          logdir=frozen_out_path,
+                          name=f"{frozen_graph_filename}.pb",
+                          as_text=False)
+
+        # Below, we assume that for multi input/output models col_start will be the same as input/output
+        # index. Example, if you have 10 inputs and each input is a simple 1-dimensional tensor, then
+        # each input must increment the col_start value, because if we always set the col_start = 0 then
+        # each Tensor will be filled from the first column and have the same value.
+
+        inputs = []
+        col_start = 0
+        for input_idx, inp in enumerate(frozen_func.inputs):
+            # print(inp.op.name)
+            # print(inp.get_shape())
+            input_dims = [1 if e is None else e for e in list(inp.get_shape())]
+            dtype = get_str_from_dtype(inp.dtype, True, input_idx)
+            inputs.append(
+                {
+                    'op_name': inp.op.name,
+                    'tensor_map': [
+                        {'idx': inp.value_index, 'dim': input_dims, 'col_start': col_start, 'data_type': dtype}]
+                })
+            col_start += flat_dim(input_dims)
+
+        outputs = []
+        col_start = 0
+        for output_idx, output in enumerate(frozen_func.outputs):
+            # print(output.op.name)
+            # print(output.get_shape())
+            output_dims = [1 if e is None else e for e in list(output.get_shape())]
+            dtype = get_str_from_dtype(output.dtype, False, output_idx)
+            outputs.append(
+                {
+                    'op_name': output.op.name,
+                    'tensor_map': [
+                        {'idx': output.value_index, 'dim': output_dims, 'col_start': col_start, 'data_type': dtype}]
+                })
+            col_start += flat_dim(output_dims)
+
+        model_info = {
+            'frozen_graph': frozen_graph_filename + '.pb',
+            'input_desc': inputs,
+            'output_desc': outputs
+        }
+
+        # print(json.dumps(model_info, indent=4, sort_keys=False))
+
+        print('Saving model description file to: ' + os.path.join(frozen_out_path, 'tf_model_desc.json'))
+        with open(os.path.join(frozen_out_path, 'tf_model_desc.json'), 'w') as json_file:
+            json.dump(model_info, json_file, indent=4, sort_keys=False)
```

Comment:
I think you can consolidate this code block as well as the write_graph block from the 2 if statements